repository/databases/oracle/onchain_transaction.go

```diff
@@ -0,0 +1,238 @@
+	tableName := appconfig.OnchainTableNameMapping["onchain_transaction"]
+
+	sql1 := fmt.Sprintf(SQL_CHECK_TABLE_EXISTED, tableName)
+	result, err := odb.Client.QueryContext(ctx, sql1, godror.FetchArraySize(1))
+	if err != nil {
+		return err
+	}
+
+	defer result.Close()
+
+	var existed int
+	for result.Next() {
+		if err := result.Scan(&existed); err == nil {
+			err := fmt.Errorf("table %s existed", tableName)
+			odb.logger.Warn(err)
+			return err
+		}
+	}
+
+	sql2 := fmt.Sprintf(sql_INIT_TABLE_ONCHAIN_TRANSACTION, tableName)
+	_, err = odb.Client.ExecContext(ctx, sql2)
+	if err != nil {
+		odb.logger.Errorf("Failed to create table %s : %v", tableName, err)
+		return err
+	}
+
+	odb.logger.Infof("table %s is created successfully", tableName)
+
+	sql3 := fmt.Sprintf(sql_INIT_INDEX_1_ONCHAIN_TRANSACTION, tableName, odb.Schema, tableName)
+	_, err = odb.Client.ExecContext(ctx, sql3)
+	if err != nil {
+		odb.logger.Errorf("Failed to index of table %s : %v", tableName, err)
+		return err
+	}
+
+	odb.logger.Infof("index of %s is created successfully", tableName)
+
+	sql4 := fmt.Sprintf(sql_INIT_INDEX_2_ONCHAIN_TRANSACTION, tableName, odb.Schema, tableName)
+	_, err = odb.Client.ExecContext(ctx, sql4)
+	if err != nil {
+		odb.logger.Errorf("Failed to index of table %s : %v", tableName, err)
+		return err
+	}
+
+	odb.logger.Infof("index of %s is created successfully", tableName)
+
+	return nil
+}
+
+func (odb *OracleDB) SaveOnchainTransaction(ctx context.Context, dbTransaction *sql.Tx, tx *types.OnchainTransactionsOracle) (int64, error) {
+	// Prepare the SQL statement
+	tableName := fmt.Sprintf("%s.%s", odb.Schema, appconfig.OnchainTableNameMapping["onchain_transaction"])
+	sql := fmt.Sprintf(sql_INSERT_ONCHAIN_TRANSACTION, tableName)
+
+	var rowsAffected int64
+
+	if dbTransaction != nil {
+		result, err := dbTransaction.ExecContext(ctx,
+			sql,
+			tx.Txhash,
+			tx.TraceNo,
+			tx.OriginTraceNo,
+			tx.SenderId,
+			tx.ReceiverId,
+			tx.FeeAccountId,
+			tx.SenderAmount,
+			tx.ReceiverAmount,
+			tx.FeeAmount,
+			tx.BlockNum,
+			tx.NetworkName,
+			tx.ChannelName,
+			tx.TxType,
+			tx.TimeOnchain,
+			tx.TimeOffchain,
+			tx.TransformTaskId,
+		)
+
+		if err != nil {
+			return 0, err
+		}
+
+		rowsAffected, _ = result.RowsAffected()
+
+	} else {
+
+		result, err := odb.Client.ExecContext(ctx,
+			sql,
+			tx.Txhash,
+			tx.TraceNo,
+			tx.OriginTraceNo,
+			tx.SenderId,
+			tx.ReceiverId,
+			tx.FeeAccountId,
+			tx.SenderAmount,
+			tx.ReceiverAmount,
+			tx.FeeAmount,
+			tx.BlockNum,
+			tx.NetworkName,
+			tx.ChannelName,
+			tx.TxType,
+			tx.TimeOnchain,
+			tx.TimeOffchain,
+			tx.TransformTaskId,
+		)
+
+		if err != nil {
+			return 0, err
+		}
+
+		rowsAffected, _ = result.RowsAffected()
+
+	}
+
+	return rowsAffected, nil
+}
+
+func (odb *OracleDB) SaveOnchainTransactionOracle(ctx context.Context, dbTransaction *sql.Tx, tx *types.OnchainTransactionsOracle) (int64, error) {
```

Comment:
Remove the duplicate function `SaveOnchainTransactionOracle` as it is identical to `SaveOnchainTransaction`. Consolidate the functionality into a single function to reduce redundancy. [medium]